python/modem-firmware-1.3+/claim_devices.py

```diff
@@ -0,0 +1,164 @@
+                     + Style.RESET_ALL) if not plain else line
+
+def hivis_style(line):
+    return ensure_lf(Fore.MAGENTA + line
+                     + Style.RESET_ALL) if not plain else line
+
+def send_style(line):
+    return ensure_lf(Fore.BLUE + line
+                     + Style.RESET_ALL) if not plain else line
+
+def error_style(line):
+    return ensure_lf(Fore.RED + line + Style.RESET_ALL) if not plain else line
+
+def bulk_claim(api_key, csv_rows):
+    csv_str = str()
+    for row in csv_rows:
+        row_str = (','.join(row) if len(row) > 1 else row[0]) + '\n'
+        csv_str += row_str
+    if args.verbose:
+        print(str(csv_str))
+    api_res = nrf_cloud_diap.bulk_claim_devices(api_key, csv_str)
+    claimed = list()
+    failed = list()
+    try:
+        json = api_res.json()
+        if 'claimedIds' in json:
+            claimed = json['claimedIds']
+        if 'failed' in json:
+            failed = json['failed']
+    except:
+        print(error_style('Error accessing json'))
+        pass
+    if api_res.status_code in {200, 201}:
+        print(local_style(f'--> Accepted; claimed {len(claimed)}:'))
+        print(local_style(f'    {claimed}'))
+    else:
+        print(error_style(f'--> Error {api_res.status_code} on {len(failed)} rows:'))
+        print(error_style(f'    {failed}'))
+        print(error_style(f'    {api_res.text}'))
+    return len(claimed)
+
+def error_exit(err_msg):
+    cleanup()
+    if err_msg:
+        sys.stderr.write(error_style(err_msg))
+        sys.stderr.write('\n')
+        sys.exit(1)
+    else:
+        sys.exit('Error... exiting.')
+
+def main():
+    global args
+    global plain
+
+    # initialize arguments
+    args = parse_args()
+    plain = args.plain
+
+    # initialize colorama
+    if is_windows:
+        init(convert = not plain)
+
+    if args.verbose:
+        print(send_style('OS detect: Linux={}, MacOS={}, Windows={}\n'.
+                          format(is_linux, is_macos, is_windows)))
+
+    print(hivis_style('\nProvisioning API URL: ' + nrf_cloud_diap.set_dev_stage(args.stage)))
+
+    try:
+        with open(args.csv) as csvfile:
+            csv_contents = csv.reader(csvfile, delimiter=',')
+
+            row_count = 0
+            pass_count = 0
+            total_rows = 0
+            bulk_prov_csv = list()
+            for row in csv_contents:
+                row_count += 1
+                total_rows += 1
+                imei = row[0]
+                uuid = row[1]
+                attest_tok = row[2]
+                date_time = row[3]
+                print(local_style(f'{row_count}. Claiming {imei}, {uuid}, with tags: "{args.provisioning_tags}"'))
+                new_row = list()
+                new_row.append(attest_tok)
+                if args.provisioning_tags is not None:
+                    new_row.append('"' + args.provisioning_tags + '"')
+                bulk_prov_csv.append(new_row)
+                if row_count == MAX_CSV_ROWS:
```

Comment:
```suggestion
if row_count == MAX_CSV_ROWS or row_count == len(csv_contents) 
```

then you can remove the duplicate block at the end.

IDK if that will actually be more readable, though... My primary motivation for suggesting this is that it is not super readable that the trailing end is outside the for loop.

Reply:
that doesn't work -- csv.reader() does not return a sequence type; it has no `len()`.

Reply:
Ah, my bad, I see why you have done it this way, then